packages/explorerkit-server/src/components/decoders/instructions.ts

```diff
@@ -0,0 +1,89 @@
+  let parsedInstruction = {
+    programId: programId.toString(),
+    encodedData: instruction.encodedData,
+    name: null,
+    decodedData: null,
+    accountKeys: instruction.accountKeys,
+  };
+
+  let parser = idls.get(programId);
+  if (parser == null) {
+    return parsedInstruction; // Short-circuit without decodedData since IDL is missing
+  }
+  let instructionParser = parser.createParser(ParserType.INSTRUCTION) as InstructionParserInterface;
+
+  // Parse the transaction
+  const decodedInstruction = instructionParser.parseInstructions(instruction.encodedData, instruction.accountKeys);
+  const decodedInstructionWithTypes = instructionParser.parseInstructions(
+    instruction.encodedData,
+    instruction.accountKeys,
+    true
+  );
+  const finalDecodedInstructionData = postProcessDecodedInstruction(
+    decodedInstruction?.data,
+    decodedInstructionWithTypes?.data
+  );
+  return {
+    programId: instruction.programId,
+    encodedData: instruction.encodedData,
+    name: decodedInstruction?.name || null,
+    decodedData: finalDecodedInstructionData || null,
+    accountKeys: instruction.accountKeys,
+  };
+}
+
+function postProcessDecodedInstruction(decodedInstructionData?: any, decodedInstructionDataWithTypes?: any): any {
+  if (!decodedInstructionData || !decodedInstructionDataWithTypes) {
+    return null;
+  }
+
+  Object.keys(decodedInstructionDataWithTypes).forEach((key) => {
+    const property = decodedInstructionDataWithTypes[key];
+    // If [[u8, X]],  base64 encode it
+    if (
+      typeof property.type === "object" &&
+      typeof property.type.vec === "object" &&
+      Array.isArray(property.type.vec.array) &&
+      property.type.vec.array[0] === "u8"
+    ) {
+      const rawValue = decodedInstructionData[key];
+      if (rawValue && Array.isArray(rawValue) && Array.isArray(rawValue[0])) {
+        decodedInstructionData[key] = rawValue.map((arr: number[]) => Buffer.from(arr).toString("base64"));
+      }
+    }
+    // If [u8, X], base64 encode it
+    if (typeof property.type === "object" && Array.isArray(property.type.array) && property.type.array[0] === "u8") {
+      // Base64 encode the byte array from decodedInstructionData
+      const byteArray = decodedInstructionData[key];
+      if (byteArray && Array.isArray(byteArray)) {
+        decodedInstructionData[key] = Buffer.from(byteArray).toString("base64");
+      }
+    }
+    // If bytes, base64 encode it
+    if (property.type === "bytes") {
+      const bytes = decodedInstructionData[key];
+      if (bytes && bytes.constructor === Uint8Array) {
+        decodedInstructionData[key] = Buffer.from(bytes).toString("base64");
+      }
+    }
+  });
+
+  return decodedInstructionData;
+}
+
+export function getProgramIds(instructionsPerTransaction: (TopLevelInstruction[] | null)[]): string[] {
+  const allProgramIds = instructionsPerTransaction
+    .filter(Boolean)
+    .flatMap((transactionInstructions) => transactionInstructions!)
```

Comment:
Is there a way of solving this on a type level without an extra iteration just to assert types? It seems silly to have TS types negatively affect runtime performance 🤷‍♂️ 

If we know they will always be defined can't we just typecast it inside existing logic?

Reply:
Technically I can do just
```
.flatMap((transactionInstructions) => transactionInstructions || [])
```

Reply:
It's no big deal but that was not what I was trying to avoid. The extra flatMap would result in an extra O(n) iteration just to assert types. 

I could be wrong but I don't think Typescrip/Node is smart enough to optimize this away. Feel free to correct me if I'm wrong though.

If the flatmap does literally nothing I would just cast the type to the correct value in the following mapping